### PR TITLE
RFR - Remote play

### PR DESCRIPTION
**Pull request by paulSanchezMarmelab**
[web_persistence.webm](https://user-images.githubusercontent.com/130989826/234777745-099dd761-ffa4-4345-978b-c18f8fe660e2.webm)

- [X] Create a PostgresSQL docker and connect from kamon server with TypeORM
- [X] Migrate needed tables on database
- [X] Create "launch game" page
- [X] Enable joining a created game from ID
- [X] Make game playable

### CODE REVIEW

File: webapp/src/game/game.entity.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import { Board } from "@kamon/core/src/board/board";
+import { GameState } from "@kamon/core/src/game/state";
+
+@Entity()
+export class Game {
+  @PrimaryGeneratedColumn()
+  id: number;
+
+  @Column("json")
+  board: Board;
```
Comment (Nitix) [inline webapp/src/game/game.entity.ts:11]: I suggest to add players ids as two columns. And use foreign keys.

**Review by Nitix — CHANGES_REQUESTED** [review]

File: webapp/src/data-source.ts
[the remark this answers is not in the document]
Comment (paulSanchezMarmelab) [reply webapp/src/data-source.ts:1]: It's used in initialization with `new DataSource`

File: webapp/src/game/game.entity.ts (hunk 1 above)
Comment (paulSanchezMarmelab) [reply webapp/src/game/game.entity.ts:11]: It was also what I planned to do,  but it's still heavily WIP since I needed to fix migrations.

File: kamon_db
Comment (Nitix) [inline kamon_db:1]: Unwanted file commited by error ?

File: package.json [hunk 2]
```diff
@@ -9,5 +9,8 @@
       "core",
       "webapp"
     ]
+  },
+  "dependencies": {
+    "@kamon/core": "workspace:^"
```
Comment (Nitix) [inline package.json:14]: This should not be here. but in the packages that import it.

File: webapp/public/app.css
Comment (Nitix) [inline webapp/public/app.css:1]: Your file has inconsistant spacing. Sometimes two spaces, sometimes one tab. (We prefer two spaces)

File: webapp/src/game/game.service.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
+import { InjectRepository } from "@nestjs/typeorm";
+import { Repository } from "typeorm";
+import { Game } from "./game.entity";
+import { initGameState, initRandomGame, Board, GameState } from "@kamon/core";
+
+export class GameService {
+  games: Repository<Game>;
```
Comment (Nitix) [inline webapp/src/game/game.service.ts:7]: I prefer gameRepository to know quickly what is by just reading the class member.
Games can also designate a list of games.

File: core/src/index.ts [hunk 4]
```diff
@@ -1,7 +1,7 @@
...
-export * from "./player";
-export * from "./randomizer";
-export * from "./tile";
+export * from "./board/board";
+export * from "./game/state";
+export * from "./graph/graph";
+export * from "./move/move";
+export * from "./player/player";
+export * from "./randomizer/randomizer";
+export * from "./tile/tile";
```
Comment (adguernier) [inline core/src/index.ts:7]: Are you sure you need this ? I added an index in every subfolder to avoid this

File: kamon_db [hunk 5]
```diff
@@ -0,0 +1 @@
+psql: error: connection to server on socket "/var/run/postgresql/.s.PGSQL.5432" failed: FATAL:  database "-" does not exist
```
Comment (adguernier) [inline kamon_db:1]: Is this should be in the repository ?

File: webapp/public/app.css [hunk 6]
```diff
@@ -77,6 +77,10 @@ body {
   display: none;
 }
 
+.board [type="submit"].show {
```
Comment (adguernier) [inline webapp/public/app.css:80]: What is the purpose ? Every tile in board are [type="submit"] and should be hidden

File: webapp/views/partials/HUD.hbs [hunk 7]
```diff
@@ -1,7 +1,7 @@
 <aside class="players-hud">
     <div class="turn-counter">
-        <h6>Turn</h6>
-        <em>{{game.gameState.turnNumber}}</em>
+        <h6>Turn n°</h6>
+        <em>{{add game.gameState.turnNumber 1}}</em>
```
Comment (Nitix) [inline webapp/views/partials/HUD.hbs:4]: It's strange that you add one to turn number here.

File: webapp/public/app.css [hunk 8]
```diff
@@ -187,4 +191,331 @@ body {
...
+	filter: progid:DXImageTransform.Microsoft.gradient( startColorstr='#82c8a0', endColorstr='#82c79e', GradientType=0 );
+}
+
+.btn-gradient.red:active 	  {background: #E35252;}
+.btn-gradient.orange:active {background: #E8601B;}
+.btn-gradient.cyan:active 	{background: #169499;}
+.btn-gradient.blue:active 	{background: #608FBF;}
+.btn-gradient.purple:active {background: #BD8EB7;}
+.btn-gradient.yellow:active {background: #DBC05B;}
+.btn-gradient.green:active  {background: #72B08E;}
```
Comment (adguernier) [inline webapp/public/app.css]: Wow ! Do we really need all of this CSS rules ? It seems to me we have one or two buttons in the app :smile:

**Review by Nitix — CHANGES_REQUESTED** [review]

File: webapp/public/app.css (hunk 6 above)
Comment (paulSanchezMarmelab) [reply webapp/public/app.css:80]: I use a submit button for the "start new game" page.

File: webapp/public/app.css (hunk 8 above)
Comment (paulSanchezMarmelab) [reply webapp/public/app.css]: Sure, but I just copy/pasted some clean css instead of writing it myself.

Comment (Nitix) [reply webapp/public/app.css]: At least add a link to the source file

File: .github/workflows/makefile_ci.yml [hunk 9]
```diff
@@ -38,16 +38,34 @@ jobs:
...
+          echo "PORT="$PORT >> .env
+          echo "RUN_MIGRATIONS="$RUN_MIGRATIONS >> .env
+          cat .env
+
+      - name: Run postgres docker
+        run: docker-compose -f "docker-compose.dev.yml" up -d --build
+
+      - name: Checking container
+        run: docker ps -a
+
```
Comment (Nitix) [inline .github/workflows/makefile_ci.yml:68]: Be careful there, you need to remove the tests run
```suggestion
      - name: Run postgres docker
        run: docker-compose -f "docker-compose.dev.yml" up -d --build

      - name: Run Webapp unit tests
        run: make unit-test-webapp

      - name: Run Webapp e2e tests
        run: make e2e-test-webapp
        env:
          POSTGRES_HOST: 127.0.0.1
          POSTGRES_PORT: 5432
          POSTGRES_USER: postgres-kamon
          POSTGRES_PASSWORD: test
          POSTGRES_DATABASE: kamon_db
```

Comment (Nitix) [reply .github/workflows/makefile_ci.yml:68]: You can also use secret like in main branch
```
      env:
        SSH_USER: ${{ secrets.AWS_SSH_USER }}
        SSH_KEY: ${{ secrets.AWS_SSH_KEY }}
        SSH_HOST: ${{ secrets.AWS_SSH_HOST }}
```

**Review by Nitix — CHANGES_REQUESTED** [review]

**Review by Nitix — APPROVED** [review]

**Pull request merged**